### DISCUSSION TITLE
ESLint language plugin for a language model.

### DISCUSSION THREAD

**Discussion by mnkiefer** [eslint/eslint#17063] [General] [answered]
With respect to the upcoming feature for ESLint Language plugins (see [RFC](https://github.com/eslint/rfcs/pull/99)), I've reconsidered some workarounds we needed to implement our plugin for the CDS language model ([CSN](https://cap.cloud.sap/docs/cds/csn?q=CSN)): [`@sap/eslint-plugin-cds`](https://www.npmjs.com/package/@sap/eslint-plugin-cds). Currently, there is only on workaround left that the RFC could not solve (generically) and I'm wondering what would be the best way to go about solving this last piece of the puzzle.

First of all, our plugin does not rely on an AST but the CDS language model which is computed from multiple files. Some of our rules work on parsed files and follow the usual ESLint flow, while other rules can't be considered for each file in isolation but only in the context of the entire model.

Hence, for the first parsed file belonging to a certain model: We compute the model, run all of the lint rules on it and cache the results. If the next file is part of that model, the lint results for that file can simply be collected from cache and be reported.

In our current approach, we have an extra (rule) handler which takes care of this (has it's own Cache).

Here, I'm wondering whether it would be possible for ESLint's `LintResultCache` (or similar) to receive results for files it has not yet seen? If not, how this could be handled more elegantly?

**Comment by nzakas** [comment]
> In our current approach, we have an extra (rule) handler which takes care of this (has it's own Cache).

Can you share this logic? It would be helpful to see what you are currently doing to determine if there's a better way to approach it.

**Reply by mnkiefer** [reply]
Yes, of course! Please refer [here](https://www.npmjs.com/package/@sap/eslint-plugin-cds?activeTab=code) to file `@sap/eslint-plugin-cds/lib/utils/createRule.js` for the following text. I've highlighted the important parts in a reduced code snippet below:
```js
module.exports = (spec) => {
  let { meta, create } = spec
  ...
  return {
    create: (context) => ({ //L34
      Program: (node) => {
        const cdscontext = extendContext(node, context, meta)
        ...   
        switch (meta.model) { // L44
          ...
          case 'B': // L55
            if (isValidFile && doRootModelChecks) {
              if (isTest || !Cache.has(`rule:${cdscontext.id}:${Cache.get('rootpath')}`)) {
                ...
                createReport(node, cdscontext, meta, create) // <- (i) run the rule (and cache report)
              } else {
                if (Cache.has(`report:${context.getFilename()}:${context.id}`)) {
                  ...
                  context.report(JSON.parse(r))  // <- (ii) collect the report (from cache) 
                }
              }
            }
            break //L72
            ...
        }
      }
    })
  }
}
```

**L34**: This is how our rules currently hook into ESLint (via a dummy Program node).

**L44**: We have two types of rules, let's call them *Type A* and *Type B*. *Type A* rules at individual files and follow the standard ESLint flow via `createReport()`. *Type B* rules on the other hand are rules that can't be considered for each file in isolation but only in the context of the entire model.

**L55-72**: Here, for a given file and *Type B* rule, we either (i) run the rule or (ii) collect the report.

Once the *ESLint Lint language plugin* becomes available and we can:
 - Register our own selectors, which get's rid of:
    - The dummy Program node (see above)
    - Our custom handler hooks (see above and **L128-164**) 
 - Get rid of *Type B* rule report recycling? 

This already simplifies our plugin tremendously, but what would be the best place for the *Type B* rule report recycling? Note, that the model computation is our rate-limiting step as these can be quite large, so once we have run all the *Type B* rules on it once, we already have the rule reports for all files belonging to that mode and no longer call parse & traverse, just report the lint errors.

Thank you for taking the time to have a look at this, I really appreciate it.

**Reply by nzakas** [reply]
Thanks for providing this! We are a bit shorthanded at the moment, so I've got a mountain of other work to tend to, but I will circle back to this soon.

**Reply by nzakas** [reply]
Just wanted to drop in and let you know that this is still on my radar. I've been moving cross country so my time has been limited, but hopefully will be settled here again in a couple of weeks and be able to circle back to this.

**Reply by nzakas** [reply]
My monthly check in to say this is still on my list. :) Just moved across the country and digging out of the backlog of work, but I haven't lost track of this.

**Comment by nzakas** [comment, marked as answer]
Okay, so finally had some time to digest this and I think the answer right now is that we don’t have a good solution for type B rules. The way ESLint is structured, it assumes each file is being evaluated in isolation and changing that would be pretty significant

On the other hand, we have had some discussion about introducing the concept of projects into ESLint. (Discussed here: https://github.com/eslint/eslint/discussions/16819) It seems like your type B rules would more naturally slide into the concept of a project, where we would have knowledge of all the files involved. 

If we had the concept of a project, then we could also have the concept of project-level rules that don’t evaluate files as individual but rather as part of the larger project. He assumption is that the user could define all the files in the project and ESLint could go from there. 

We are still a ways away from determining what a project is for ESLint, but does what I’m describing seem like a potential fit?

**Reply by mnkiefer** [reply]
Yes, that would be ideal. I will keep watching the ["Project-aware ESLint" discussion](https://github.com/eslint/eslint/discussions/16819) for any updates. 👀

Thanks again for your help!
